### PR TITLE
[test_fgnhg] Add test to validate FG ECMP packet flow when neighbors are not resolved

### PR DESCRIPTION
**Pull request by anish-n**
### Description of PR

Summary:
Test case enhancement for newly added capability in Fine Grained ECMP to handle packet flow upon no ECMP next-hops being neighbor resolved. 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

### Back port request
- [ ] 201911

#### What is the motivation for this PR?
Test case enhancement for newly added capability in Fine Grained ECMP to handle packet flow upon no ECMP next-hops being neighbor resolved

#### How did you do it?
The test will start off with no neighbors resolved and then send packets to the Fine Grained ECMP destination, we will validate that the packets did make it to the endpoint, and validate that the neighbor entry ended up getting resolved via the normal packet flow.

#### How did you verify/test it?
Ran the pytest with PTF

### CODE REVIEW

File: ansible/roles/test/files/ptftests/fg_ecmp_test.py [hunk 1]
```diff
@@ -199,14 +262,33 @@ def fg_ecmp(self):
...
+                else:
+                    in_port = self.net_ports[0]
+                (port_idx, _) = self.send_rcv_ip_pkt_lossy(
+                    in_port, src_port, dst_port, src_ip, dst_ip, self.serv_ports, ipv4)
+                if port_idx == -1:
+                    total_num_pkts_lost = total_num_pkts_lost + 1
+
+            self.log("Number of lost packets were: " + str(total_num_pkts_lost))
+            # Ensure less than 1% packet loss
+            assert (total_num_pkts_lost < (0.01 * self.num_flows))
```
Comment (nazariig) [inline ansible/roles/test/files/ptftests/fg_ecmp_test.py:289]: @anish-n can you please add a constant for this (0.01) value?

File: ansible/roles/test/files/ptftests/fg_ecmp_test.py [hunk 2]
```diff
@@ -158,7 +215,12 @@ def fg_ecmp(self):
             with open(PERSIST_MAP, 'w'): pass
```
Comment (nazariig) [inline ansible/roles/test/files/ptftests/fg_ecmp_test.py:215]: @anish-n what's the purpose of this code?

File: tests/ecmp/test_fgnhg.py [hunk 3]
```diff
@@ -126,6 +126,37 @@ def setup_neighbors(duthost, ptfhost, ip_to_port):
...
+
+    ptfhost.copy(src='/tmp/from_t1.json', dest='/tmp/from_t1.json')
+
+    extra_vars = {
+            'arp_responder_args': ''
+    }
+
+    ptfhost.host.options['variable_manager'].extra_vars.update(extra_vars)
+    ptfhost.template(src='templates/arp_responder.conf.j2', dest='/tmp')
+    ptfhost.command("cp /tmp/arp_responder.conf.j2 /etc/supervisor/conf.d/arp_responder.conf")
```
Comment (nazariig) [inline tests/ecmp/test_fgnhg.py:151]: @anish-n a typo? `cp /tmp/arp_responder.conf /etc/supervisor/conf.d/arp_responder.conf`

File: ansible/roles/test/files/ptftests/fg_ecmp_test.py [hunk 4]
```diff
@@ -29,6 +29,63 @@
 
 PERSIST_MAP = '/tmp/fg_ecmp_persist_map.json'
 
+def verify_packet_warm(test, pkt, port, device_number=0, timeout=None, n_timeout=None):
```
[the remark this answers is not in the document]
Comment (anish-n) [reply ansible/roles/test/files/ptftests/fg_ecmp_test.py:32]: This is a standard definition in p4lang/ptf, want to keep the default definitions intact

File: ansible/roles/test/files/ptftests/fg_ecmp_test.py [hunk 5]
```diff
@@ -158,7 +215,12 @@ def fg_ecmp(self):
             with open(PERSIST_MAP, 'w'): pass
         else:
             with open(PERSIST_MAP) as fp:
-                tuple_to_port_map = json.load(fp)
+                try:
+                    tuple_to_port_map = json.load(fp)
+                except ValueError:
+                    print 'Decoding JSON failed for persist map'
+                    if not self.test_case == 'verify_packets_received':
+                        assert False
```
[the remark this answers is not in the document]
Comment (anish-n) [reply ansible/roles/test/files/ptftests/fg_ecmp_test.py]: ```fail``` is not used in ptf test, will continue using assert

**Review by nazariig — APPROVED** [review]

**Review by prsunny — APPROVED** [review]

**Comment by bingwang-ms** [comment]
@anish-n We are seeing errors in `fg_ecmp_test.py` possibly caused by typo
`
"======================================================================", 
        "ERROR: fg_ecmp_test.FgEcmpTest", 
        "----------------------------------------------------------------------", 
        "Traceback (most recent call last):", 
        "  File \"ptftests/fg_ecmp_test.py\", line 549, in runTest", 
        "    self.fg_ecmp()", 
        "  File \"ptftests/fg_ecmp_test.py\", line 277, in fg_ecmp", 
        "    src_ip = str(base_ip + i)", 
        "NameError: global name 'base_ip' is not defined", 
        "", 
        "----------------------------------------------------------------------", 
        "Ran 1 test in 3.008s", 
        "", 
        "FAILED (errors=1)"
`
Can you help take a look? Thanks

**Comment by anish-n** [comment]
> @anish-n We are seeing errors in `fg_ecmp_test.py` possibly caused by typo `"======================================================================", "ERROR: fg_ecmp_test.FgEcmpTest", "----------------------------------------------------------------------", "Traceback (most recent call last):", " File \"ptftests/fg_ecmp_test.py\", line 549, in runTest", " self.fg_ecmp()", " File \"ptftests/fg_ecmp_test.py\", line 277, in fg_ecmp", " src_ip = str(base_ip + i)", "NameError: global name 'base_ip' is not defined", "", "----------------------------------------------------------------------", "Ran 1 test in 3.008s", "", "FAILED (errors=1)"` Can you help take a look? Thanks

Thanks for letting me know @bingwang-ms, will fix it today and send a PR.

**Pull request merged**